### ISSUE TITLE
Matplotlib error for Force graph

### ISSUE THREAD

**Issue by MES-physics** [libAtoms/QUIP#513]
Hi,
After many trials and experiments from ideas online, I keep getting this error for the plotting code:
Traceback (most recent call last):
  File "./MATPLOTnew.sh", line 104, in <module>
    force_plot('C3909evenhalf.xyz', 'quip_trainC3909Sehalf.xyz', ax_list[2], 'Force on training data')
  File "./MATPLOTnew.sh", line 83, in force_plot
    flim = (for_limits.min() - 1, for_limits.max() + 1)
  File "/home/m/.local/lib/python3.6/site-packages/numpy/core/_methods.py", line 43, in _amin
    return umr_minimum(a, axis, None, out, keepdims, initial, where)
ValueError: zero-size array to reduction operation minimum which has no identity

Is there a usual fix for this?   Here is my code:  
[MATPLOTnew.txt](https://github.com/libAtoms/QUIP/files/10116748/MATPLOTnew.txt)

**Comment by bernstei** [comment]
Looks like `for_limits` has dimension zero, so you can't take its minimum or maximum.  You need to figure out why it has dimension 0, which would mean that `ener_in` and `ener_out` have dimension zero, which would mean that both the files you're reading have zero configurations.

**Comment by MES-physics** [comment]
OK, I looked at this and my files have thousands of configurations with 50-94MBs and the file names are correct. The energy plots work separately too. It is the force plots that don't.
So now I'm wondering if the files got corrupted with respect to forces somehow in the quip command to make the quip_train*.xyz and quip_validate*.xyz files.  Here is the **start** of the format for the quip_train*.xyz type of file attached. The real file is about 94MBs.  I had this working for another set (the Carbon_GAP_20 as well as another of my own) but can't see any problems.  Anything wrong in this format?
[quip_trainfile.txt](https://github.com/libAtoms/QUIP/files/10118282/quip_trainfile.txt)

Eternally grateful.

**Comment by bernstei** [comment]
You need to debug the script that constructs `for_limits`.  Add some statements that print the length of the lists that get concatenated (`in_force`, `out_force`) - are they zero, like I suspect?  If so, add some statements that print things when those lists get constructed, and figure out why they're empty lists.

**Comment by MES-physics** [comment]
I was able to find out by the print statements that the j loop seems not working to gather the forces; right after it, the ```in_force``` and ```out_force``` are still empty.
Atoms in and out are good.  Symbols will print. Energies are good.  I don't see where to go from here.  Maybe the symbols are somehow out of sync with the force values in this file.  I'm wondering how they are supposed to be arranged then.
```
       for j, sym in enumerate(sym_all):
            if sym in symbol:                
                in_force.append(at_in.get_forces()[j])
                #out_force.append(at_out.get_forces()[j]) \
                out_force.append(at_out.arrays['force'][j]) # because QUIP and ASE use different names
    # convert to np arrays, much easier to work with
    print(in_force)    # < these print [] (empty )
    print(out_force)  # < these print [] (empty) 
    in_force = np.array(in_force)
    out_force = np.array(out_force)
```

**Comment by bernstei** [comment]
This is a python question, nothing specific to libAtoms/QUIP, and it's just debugging.   Go back one step. If those lists are empty, the loops aren't looping, otherwise the `in_force.append()` would have to be adding something (if it got no actual forces). What's in `sym_all`? Add print statements in every loop until you figure out which loop isn't doing any iterations.

**Comment by MES-physics** [comment]
Yes, symbols print as I tried above, the sym_all prints  'C', 'C', 'C', 'C', 'C', etc etc.
And the outer loops print.   It is the innermost loop trying to do appending that doesn't work.  Sorry, python is new to me too, I'm trying to learn everything at once.
```
       sym_all = at_in.get_chemical_symbols()
        print(sym_all)   #prints multiple 'C' 'C' etc.
        # add force for each atom
        for j, sym in enumerate(sym_all):
            print(in_force)   #prints multiple [] [] [] [] etc.
            if sym in symbol:                
                in_force.append(at_in.get_forces()[j])
                print(in_force)  # does not print anything
                #out_force.append(at_out.get_forces()[j]) \
                out_force.append(at_out.arrays['force'][j]) # because QUIP and ASE use different names
                print(out_force) #does not print anything
```

**Comment by bernstei** [comment]
If the prints you indicate are printing nothing are printing nothing (not empty list), it must not be executing the code inside the `if` block, so it must be that the `if sym in symbol:` is always False.  I'd suggest printing `sym` and `symbol` right before the `if` line to try to figure out why that test is apparently always evaluating to `False`.

**Comment by MES-physics** [comment]
Ok thank you, here are those results:
```        
            for j, sym in enumerate(sym_all):
            print(sym)  # prints a column of C  C  C  C etc.  down the left side of screen
            print(symbol) # prints a column of ' Force on training data ' down the left side of the screen 
            if sym in symbol:                
                in_force.append(at_in.get_forces()[j])
                #out_force.append(at_out.get_forces()[j]) \
                out_force.append(at_out.arrays['force'][j]) # because QUIP and ASE use different names
```
I see that ```symbol``` above in the program is defined as C though.
```
def force_plot(in_file, out_file, ax, symbol='C', title='Plot of force'):
```

**Comment by bernstei** [comment]
That's just the default value. When you _call_ `force_plot(...)` you can pass a different value that will override the default.

**Comment by MES-physics** [comment]
OK, I tried another experiment.  Here is a script that works correctly on another element file, and what the print statements do in every case.  I did every print statement individually to find out what the output should be when the script works.
```
 for j, sym in enumerate(sym_all):
#           print(symbol) # prints As (arsenic) column down screen
#           print(sym)   # prints As column down screen
            if sym in symbol:
                in_force.append(at_in.get_forces()[j])
#               print(symbol) # prints As column down screen
#               print(sym)    # prints As column down screen
                #out_force.append(at_out.get_forces()[j]) \
                out_force.append(at_out.arrays['force'][j]) # because QUIP and ASE use different names
#               print(sym) # prints As column down screen
#               print(symbol) # prints As column down screen
    # convert to np arrays, much easier to work with
#   print(in_force)   #  prints arrays of forces
 #  print(out_force)  #  prints arrays of forces
```

**Comment by bernstei** [comment]
I don't know what you did to get it to work, but you already know that in your real run `symbol` doesn't contain `C`, since you printed its value. When you are calling `force_plot(....)`, you're passing a value for the `symbol` argument (from your file attached above):
```
force_plot('C3909evenhalf.xyz', 'quip_trainC3909Sehalf.xyz', ax_list[2], 'Force on training data')
```
It's running with `symbol = 'Force on training data'`, exactly like you asked it to when you called `force_plot` with these arguments.

**Comment by MES-physics** [comment]
Oh my, yes, I found the problem!  My number of parameters in calling ```force_plot``` did not match the number of them in the definition.  The element needed to be added before 'force on training data' like in the definition, to pick up the values it seems. I've only worked on some C and C++ in the past and now getting used to python. I'm learning fast every day, don't worry I'll try not to bother you much. Sometimes talking a bit works wonders. Grad students these days are too isolated if you've heard. 
Thank you very much, you're a great help to me.
